[MQTT] Update last activity timestamp only when data was sent to broker

TD-er opened the pull request:
The PubSubClient does keep track of the last incoming and outgoing activity.
These timestamps are used to determine when a keep-alive packet has to be sent.
However, there was no check for the return value of writing data to the broker.
So the client could mistakenly assume it was sending keep-alive packets while the broker was not receiving any.
This can lead to strange situations where the connected state was not the same on the client side and the broker side.

TD-er committed:
[PubSubClient] Fix check for domain string

TD-er committed:
[PubSubClient] Only update lastOutActivity when write is successful

TD-er committed:
[MQTT] Improve update MQTT client connected status

Checking the MQTT connected status can alter connected status, so update connected status when it could have changed.

v-a-d-e-r commented:
Another bug you were building in.....
        // Leave room in the buffer for header and variable length field
        uint16_t length = MQTT_MAX_HEADER_SIZE;
        length = writeString(topic,buffer,length);
        uint16_t i;
to:
        // Leave room in the buffer for header and variable length field
        uint16_t length = writeString(topic,buffer,length);
        uint16_t i;

MQTT will stop working!! It should be:

        // Leave room in the buffer for header and variable length field
        uint16_t length = writeString(topic,buffer,MQTT_MAX_HEADER_SIZE);
        uint16_t i;

TD-er commented:
Thanks.
Missed that one and indeed that's something I changed.
The test build I made was indeed not working.

TD-er committed:
[pubsubclient]  Fix bug in publish

Change I made which was incorrect.
As spotted here: https://github.com/letscontrolit/ESPEasy/pull/2728#issuecomment-552925267

v-a-d-e-r commented:
You are welcome ;-)

TD-er commented:
I made a [test version - ESPEasy_mega-20191108-36-PR_2728.zip](https://www.dropbox.com/s/g2xk0884j02dx0d/ESPEasy_mega-20191108-36-PR_2728.zip?dl=0)

Would be nice if this can be tested by someone with unstable MQTT.

TD-er committed:
[pubSubClient]  Fix signed/unsigned compare warning

TD-er committed:
[SD] Allow to compile with FEATURE_SD defined (#2700)

However it is unusable, see https://github.com/esp8266/Arduino/issues/6767

TD-er committed:
[PIO] Ignore SD/SDFS libraries to reduce build size

See https://github.com/esp8266/Arduino/issues/6760

TD-er committed:
[C002] Allow heatpump commands for new plugin P088

Plugin was recently moved from playground to main repository, but the commands sent from Domoticz MQTT were not allowed for the new plugin ID.
